hpack/src/main/java/com/twitter/hpack/Decoder.java

```diff
@@ -460,42 +472,54 @@ private void readName(int index) throws IOException {
     }
   }
 
-  private void indexHeader(int index, HeaderListener headerListener) throws IOException {
+  private void indexHeader(int index, ExtendedHeaderListener headerListener) throws IOException {
     if (index <= StaticTable.length) {
       HeaderField headerField = StaticTable.getEntry(index);
-      addHeader(headerListener, headerField.name, headerField.value, false);
+      // The static table has a set of constant annotations so don't allow them to mutate
+      addHeader(headerListener, headerField.name, headerField.value,
+          staticTableAnnotations[index - 1], false);
     } else if (index - StaticTable.length <= dynamicTable.length()) {
       HeaderField headerField = dynamicTable.getEntry(index - StaticTable.length);
-      addHeader(headerListener, headerField.name, headerField.value, false);
+      addHeader(headerListener, headerField.name, headerField.value, headerField.annotations,
+          false);
     } else {
       throw ILLEGAL_INDEX_VALUE;
     }
   }
 
-  private void insertHeader(HeaderListener headerListener, byte[] name, byte[] value, IndexType indexType) {
-    addHeader(headerListener, name, value, indexType == IndexType.NEVER);
-
-    switch (indexType) {
-      case NONE:
-      case NEVER:
-        break;
-
-      case INCREMENTAL:
-        dynamicTable.add(new HeaderField(name, value));
-        break;
-
-      default:
-        throw new IllegalStateException("should not reach here");
+  private void insertHeader(ExtendedHeaderListener headerListener, byte[] name,
+                            byte[] value, IndexType indexType) {
+    try {
+      addHeader(headerListener, name, value, tmpAnnotations, indexType == IndexType.NEVER);
+      switch (indexType) {
+        case NONE:
+        case NEVER:
+          break;
+
+        case INCREMENTAL:
+          dynamicTable.add(HeaderField.forReceivedHeader(name, value, tmpAnnotations));
+          break;
+
+        default:
+          throw new IllegalStateException("should not reach here");
+      }
+    } finally {
+      // The header was annotated so need to create a new slot to avoid overwriting the
```

Comment:
probably need to always create a new array whenever we pass it to the consumer -- otherwise they could store the value and later mutate multiple header entries